File: Heap/kth/kthSmallest2SortedArr.java
//solution logk https://www.programcreek.com/2012/12/leetcode-median-of-two-sorted-arrays-java/
public int kSmallest2SortedArr(int[] a, int[] b, int k){
  return helper(a, 0, b, 0, k);
}
private int helper(int[] a, int aleft, int[] b, int bleft, int k){
  //base case
  if(aleft >= a.length)
    return b[bleft+k/2-1];
  if(bleft >= b.length)
    return a[aleft+k/2-1];
  if(k == 1)
    return Math.min(a[aleft], b[bleft]);
  int ak = aleft+k/2-1 < a.length ? a[aleft+k/2-1]:Integer.MAX_VALUE;
  int bk = bleft+k/2-1 < b.length ? b[bleft+k/2-1]:Integer.MAX_VALUE;
  if (ak < bk){
    return helper(a, aleft+k/2, b, bleft, k-k/2);
  }
  return helper(a, aleft, b, bleft+k/2, k-k/2);
}



//solution k https://www.geeksforgeeks.org/median-of-two-sorted-arrays/

File: Heap/kth/kthSmallest2SortedArr.py
# O(lgn) run time
def kthSmallest2SortedArrays(A, B, k):
    return self.helper(A, 0, B, 0, k)

def helper(self, A, aleft, B, bleft, k):
    if aleft == len(A):
        return B[bleft + k - 1]
    if bleft == len(B):
        return A[aleft + k - 1]
    if k == 1:
        return min(A[aleft], B[bleft])
    ak = A[aleft + k//2 - 1] if aleft + k//2 - 1 < len(A) else float('inf')
    bk = B[bleft + k//2 - 1] if bleft + k//2 - 1 < len(B) else float('inf')
    if ak < bk:
        return self.helper(A, aleft + k//2, B, bleft, k - k//2)
    else:
        return self.helper(B, aleft, B, bleft + k//2, k - k //2)
